File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsDir.java
/**
 * @file
 * @brief Vfs directory object interface
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs;

import java.io.IOException;

import app.zxtune.utils.ProgressCallback;

public interface VfsDir extends VfsObject {

  /**
   * Directory content enumerating callback
   */
  abstract class Visitor implements ProgressCallback {

    /**
     * Called on visited directory
     */
    public abstract void onDir(VfsDir dir);

    /**
     * Called on visited file
     */
    public abstract void onFile(VfsFile file);

    /**
     * Called when items count is known (at any moment, maybe approximate)
     */
    public void onItemsCount(int count) {}

    @Override
    public void onProgressUpdate(int done, int total) {}
  }

  /**
   * Enumerate directory content
   *
   * @param visitor Callback
   */
  void enumerate(Visitor visitor) throws IOException;
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsDir.kt
/**
 * @file
 * @brief Vfs directory object interface
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs

import app.zxtune.utils.ProgressCallback
import java.io.IOException

interface VfsDir : VfsObject {
    /**
     * Directory content enumerating callback
     */
    // TODO: make interface
    abstract class Visitor : ProgressCallback {
        /**
         * Called on visited directory
         */
        abstract fun onDir(dir: VfsDir)

        /**
         * Called on visited file
         */
        abstract fun onFile(file: VfsFile)

        /**
         * Called when items count is known (at any moment, maybe approximate)
         */
        open fun onItemsCount(count: Int) {}

        override fun onProgressUpdate(done: Int, total: Int) {}
    }

    /**
     * Enumerate directory content
     *
     * @param visitor Callback
     */
    @Throws(IOException::class)
    fun enumerate(visitor: Visitor)
}
